### DISCUSSION TITLE
A target for Cypress

### DISCUSSION THREAD

**Discussion by 7tupel** [thheller/shadow-cljs#1026] [Ideas]
Hi everyone,

I recently started using shadow-cljs in favor to other options to compile my cljs projects. One reason was to be able to use storybook for (react) components. Another thing I would like to do is use cypress for e2e and component testing.

Now this might seem not an issue for shadow-cljs but please let me explain my rationale here:

Cypress is a JS testing framework for e2e and recently also component tests. Right now the only way I found how to use cypress with Clojurescript (and shadow-cljs) is the [cypress-clojurescript-preprocessor](https://github.com/viesti/cypress-clojurescript-preprocessor). What this does is basically generate a new build configuration for each test and compile clojurescript into JS bundles. For each test file a new js bundle gets generated which is then loadable by cypress.

This approach is okeh-ish but might get problematic when there will be many hundreds(?) of tests files. And its not very elegant.

A more elegant way I believe would be to have a build target in shadow-cljs that will analyze the Clojurescript modules following a `:ns-expression` like i.e. `"-test$"`, generate a JS test file for each module that can then be imported by cypress. This is exactly not how the current `:browser-test` target works which creates a single bundle with all tests (and other code).

This approach would be similar to how I currently compile the Clojurescript code into JS to be usable in storybook: here I simply use the `:node-module` target which will then not generate a bundle but independent JS files for each module. Using this target does unfortunately not work with cypress for reasons not 100% clear to me.

I would like to use this post to get feedback if anyone else has tried something like this and if this could be an interesting thing for other developers who would like to use shadow-cljs with cypress.

P.S. Here is a [discussion about the rationale](https://github.com/viesti/cypress-clojurescript-preprocessor/issues/12) in the preprocessor linked above.

**Comment by thheller** [comment]
First of: Target implementations do not need to be part of shadow-cljs. They can be developed as a library or bundled with shadow-cljs. I'm happy to help out regardless.

However, I just don't have the slightest clue what cypress does or requires. Generating some extra files is trivial, generating a full new build definitely seems like overkill.

Is there some documentation that explains how cypress builds and runs the code? is it node? does it run webpack or similar? Does it need code organized in specific ways, apart from the `cypress/integration` dir?

**Reply by 7tupel** [reply]
As far as I can tell cypress has both a node process running that controls the tests and a webbrowser instance (Cypress uses a real browser to run tests which is different from many other testing libraries like jest). For component tests it will in the end open a browser, load a webapp generated by the node process and run the code that is to be tested within the browser. Cypress can use webpack internally.

As far as I have gotten is the following:

Compiling the project with the target `:npm-module` seems to work fine first, cypress finds the tests but once I go to the component test using the cypress webui generated by cypress I get the error 
```
process is not defined
```
This makes of cause sense because inside the browser there should be no `process` module. As far as I can see the only occurrence where `process` is used is in the compiler generated `cljs_env.js` file. 

One idea I have would be injecting a fake process object to mock the single call on process in that file. Not sure though if this will work as expected using cypress.

Interestingly is one thing: i also use shadow-cljs with the `:npm-module` target to build single files from my components to use them in storybook. Storybook also uses a node process running a local server that takes the components into a sandbox and runs a webui used to interact with the components where the component code gets executed in the browser. Not errors complaining about `process` here.

**Reply by 7tupel** [reply]
Maybe there is an even nicer way to do this from shadow-cljs using a custom library.

**Reply by thheller** [reply]
Ignoring the build issues for a second. How would things ideally look on the code side? I personally would try to ignore the whole cypress setup as much as possible and instead stick with regular CLJS project structure. So, I would create files in `src/test/my/app/e2e/foo.cljs` and then

```
(ns my.app.e2e.foo)

(js/describe "example"
    (js/it "does nothing"
      (fn []
        (js/console.log "well it logs?")))))
```

This would then yield a `cypress/e2e/foo.cy.js` file that shows up in the UI properly?

I do however see several issues with this since this always runs whenever this file is loaded. Meaning it is not loadable in the REPL with these side-effects.

An alternative would be to use metadata and break from model the cypress sort of imposes via folders/files.

So, instead you could write

```
(ns my.app.e2e.foo)

(defn my-test
  {:cypress/test "1-getting-started/todo"}
  []
  (js/describe "example"
    (js/it "does nothing"
      (fn []
        (js/console.log "well it logs?")))))

(defn my-other-test
  {:cypress/test "1-getting-started/other"}
  []
  (js/describe "example"
    (js/it "does nothing"
      (fn []
        (js/console.log "well it also logs?")))))
```

In this way the compiler would look for the `:cypress/test` (or maybe `:cypress/spec`?) metadata and generate files according to that.

So, this way you'd end up with `cypress/e2e/1-getting-started/todo.cy.js` and `cypress/e2e/1-getting-started/other.cy.js` files instead of one.

I have never used cypress in any project so I don't have a clue how relevant this even is. Would it be desirable to declare multiple specs in one file? The `js/describe` etc things could look a little nicer with macros of course.

**Reply by thheller** [reply]
Does it really make sense to use CLJS to write these specs?

I mean the default syntax for testing assertions and all that looks horrific in CLJS no? Just going over some of the examples they provide and translating those to CLJS looks horrible and rather interop heavy?

**Reply by 7tupel** [reply]
You can structure the tests for cypress in different ways. It makes sense to have some directory/namespace structure like the one you described where you have a `my.app.e2e.foo` module that describes either a single integration test or a whole suite of them. It doesn't really matter if you describe a single testsuite in one module or more than one (its up to how the developer would like to structure the tests).

The same goes for component tests. Personally I like to have a component test directly next to the component itself i.e. like `my.app.components.foo-cypress` but you could put them into an extra directory like `my.app.component-tests.foo`. Again the module can either describe a single suite of tests or many. Each suite can of cause have any number of test cases.

Loading the file from the repl might indeed cause unpredictable side effects. But on the other hand it wouldn't make much sense to load the modules with the cypress tests from the repl because cypress is intended to run tests within a real web browser. This is different from the Clojure built-in unit testing. (it is of cause possible to also have standard unit tests using the Clojure toolchain in addition to cypress. but the should reside in different modules).

I could of cause write the cypress tests in plain JS which would potentially avoid using a lot of ugly interop code in Clojurescript. (but this would not resolve the original problem with the node `process` requirement as this does not reside in the code generated specifically for the test). Another approach would be using a nicer wrapper like `mocha-latte` and a bit of (yet to be written) lib to call cypress specific functions from Clojurescript in a nice way.

Example for a component test:
```Clojure
(ns my.app.component-cypress
  ""
  (:require-macros [latte.core :refer [describe it]]))

(describe "Component Test Suite"
  (it "This does nothing Hooray" []
    true)
  (it "Not doing enything either" []
    true))

(describe "Another Component Test Suite"
  (it "Really stupid this one" []
    true))
```

Using Cypress directly from Clojurescript would then look something like this:
```Clojure
(describe "Testing a component with cypress"
  (it "Some Test" []
    (mount MyComponent)
    (as-> js/cy cy
      (.get cy "h1")
      (.contains cy "title"))))
```

That is indeed not very nice code but its at least a starting point from which a mode Clojure friendly wrapper could be created for cypress (of cause that is not something related to shadow-cljs)

**Reply by 7tupel** [reply]
As for the compilation issue, even when writing cypress tests in JS I would still not be able to just use the components written with Clojurescript because the generated JS code of the component (i.e. `my.app.component.js`) will still have a `var $CLJS = require("./cljs_env");` and the `cljs_env` module will still have a `var CLJS_GLOBAL = process.browser ? (typeof(window) != 'undefined' ? window : self) : global;` line at the beginning which will cause the cypress test server to raise the error about the missing `process`.

Is there a way how to change the way shadow-cljs generates this file (i.e. using a post-processing function or plugin)? That I guess might be leaner instead of having a whole new compile target?

Sorry I'm really not familiar with shadow-cljs yet because I have been using figwheel until now...

**Reply by thheller** [reply]
I'm just trying to determine what is actually needed here since I have never used cypress. Your feedback covered most of the things I guess. I have only looked at the `e2e` portion so far which doesn't actually seem to need any access to the "other" CLJS code whatsoever? It just runs directly against the generated DOM? I assume you are referencing components tests for the other issue?

It would help if you create a sample repo that is **not** working for you. Seeing the problem in a reproducible way actually helps a lot in figuring out what is needed. Build issues I can fix.

**Reply by 7tupel** [reply]
Yes I think that for e2e tests there should not be any problems with the compilation because I would just start the webapp and then run the Cypress process with the tests against the (local) server that provides the webapp. Cypress would only need the testcase itself but not the component or anything else from the app. Not sure though if that would still cause issues because the compiled JS file might still require the `cljs_env` file which again tries to access `process`. But I believe the e2e tests will run in a node process and not in the browser and therefore should not throw the error. I have not really tried the e2e tests yet.

The primary issue I'm having is still with the component tests.

I just invited you to my repo that I use to play with storybook, cypress and shadow-cljs. Maybe this can help. All changes are on the _development_ branch.

**Reply by thheller** [reply]
Please stop worrying about issues with `cljs_env`. I can fix those very easily. I just need to understand what is actually needed and if a custom `:target` would make things actually easier or if the existing ones would cover it fine.

`process` not being available is `webpack` complaining. Not `node`. The `e2e` tests run in an iframe in the Browser. It could be fixed with some extra webpack config but I can also just get rid of the reference completely.

I'll look into the component tests a little more. Never tried those once yet.

**Reply by thheller** [reply]
You can try `shadow-cljs` version `2.19.6` and setting `:runtime :browser` in the `:npm-module` build config. That'll no longer reference `process` and seems to just work.

I still think the entire setup is a little less than ideal given that webpack technically isn't needed. However, given that you can only pick officially supported bundlers/frameworks for this whole thing there isn't much of a choice.

**Reply by 7tupel** [reply]
Thanks for the change but it doesn't seem to work or I'm doing something wrong. Just to be sure I didn't misunderstood you: I changed the build configuration for cypress like
```edn
:cy-component {:target          :npm-module
               :runtime         :browser
               :ns-regexp       "-cypress$"
               :output-dir      "resources/cy-component/compiled"}
```

Apart from that I did the following steps:
- Remove `node_modules` directory and `yarn.lock` file
- change the version of the `shadow-cljs` dependency to `2.19.6`
- freshly install the node dependencies with yarn
- clean the build folder where the compiled JS code will be put
- start a new build
- run cypress

The generated JS code still uses the `process` variable.

**Reply by thheller** [reply]
You didn't actually bump shadow-cljs. You are using `deps.edn` so you need to update it there. You also don't need to remove the `node_modules` folder or the `.lock` file. Actually you are supposed to keep that.

**Reply by 7tupel** [reply]
well, just like I said, it might be a stupid mistake at my end ^^ 

Thanks a lot, it seems to work so far. I'll try around a bit.
